### ISSUE TITLE
Include() Appended to Query is Dependent on Order of Joins

### ISSUE THREAD

**Issue by brent-russell** [aspnet/EntityFramework#3091]
Issue encountered in Beta7.

When a query contains multiple joins, and an .Include() is appended to the query, the child entities are only included when the DbSet with the Include statement is the initial (first) from clause. In this specific case, the include is also a self referential parent/child relation.

Here's the example, which is probably easier to read through.

```C#
// Topics is first DbSet in from clause - include works
var query = from t in this.DbContext.Topics
			join i in this.DbContext.ItemTopics
				on t.Id equals i.TopicId
			join ci in this.DbContext.CategoryItems
				on i.RootItemId equals ci.RootItemId
			where ci.CategoryId == categoryId
			select t;

// Topics.Children is ICollection<Topic> where Topic.ParentTopicId = Topic.Id
query  = query.Include(t => t.Children);
// children are included
```

```C#
// Topics is NOT first DbSet in from clause - include does not work
var query = from ci in this.DbContext.CategoryItems
			join i in this.DbContext.ItemTopics
				on ci.RootItemId equals i.RootItemId
			join t in this.DbContext.Topics
				on i.TopicId equals t.Id
			where ci.CategoryId == categoryId
			select t;

// Topics.Children is ICollection<Topic> where Topic.ParentTopicId = Topic.Id
query  = query.Include(t => t.Children);
// children are NOT included
```

Edit - Added models:
```C#
public class Item
{
	public int Id { get; set; }
	public int? RootItemId { get; set; }	// self referential to Item.Id
	public Item RootItem { get; set; }
}

public class Topic
{
	public int Id { get; set; }
	public int? ParentTopicId { get; set; }	// self referential to Topic.Id
	public Topic ParentTopic { get; set; }
	public ICollection<Topic> Children { get; set; } = new List<Topic>();
}

public class ItemTopic
{
	public int Id { get; set; }
	public int RootItemId { get; set; }	// reference to Item.Id
	public int TopicId { get; set; }	// reference to Topic.Id
	public Item RootItem { get; set; }
	public Topic Topic { get; set; }
}

public class CategoryItem
{
	public int Id { get; set; }
	public int CategoryId { get; set; }
	public int RootItemId { get; set; }	// reference to Item.Id
	public Category Category { get; set; }
	public Item RootItem { get; set; }
}
```

Edit - OnModelCreating:
```C#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
	base.OnModelCreating(modelBuilder);

	modelBuilder.UseSqlServerIdentityColumns();

	modelBuilder.Entity<CategoryItem>(
		e =>
		{
			e.Property(p => p.Id).HasColumnName("CategoryItemId");
		}
	);

	modelBuilder.Entity<Item>(
		e =>
		{
			e.Property(p => p.Id).HasColumnName("ItemId");
		}
	);

	modelBuilder.Entity<ItemTopic>(
		e =>
		{
			e.Property(p => p.Id).HasColumnName("ItemTopicId");
		}
	);

	modelBuilder.Entity<Topic>(
		e =>
		{
			e.Property(p => p.Id).HasColumnName("TopicId");
		}
	);
```

**Comment by divega** [comment]
@anpete have you seen this one?

**Comment by maumar** [comment]
@brent-russell could you provide a model you used in this scenario? I tried to reverse engineer the model based on the query, but I'm seeing Children being included in both cases.

**Comment by brent-russell** [comment]
@maumar added models used in query to original comment.

**Comment by maumar** [comment]
@brent-russel thanks! Just in case, could you also add the contents of your OnModelCreating?

**Comment by brent-russell** [comment]
@maumar added OnModelCreating snippet. The Id properties get mapped to more explicitly named database column names.

**Comment by maumar** [comment]
fixed in 357102d51c9c5e9283a3b0ea00b30d2a5ea212f4
